lib/expand.js

```diff
@@ -225,6 +227,102 @@ class Expander {
     }
   }
 
+  expandHierarchy(element) {
+    const setCurrentElementAsOriginalModifier = (value) => {
+      if (value.options) {
+        value.options.forEach(v => setCurrentElementAsOriginalModifier(v));
+      } else {
+        value.constraints.forEach(c => c.lastModifiedBy = element.identifier);
+      }
+    }
+
+    if (element.basedOn.length == 0) {
+      if (element.value) setCurrentElementAsOriginalModifier(element.value);
+      if (element.fields) element.fields.forEach(f => setCurrentElementAsOriginalModifier(f));
+
+    } else {
+      for (const baseID of element.basedOn) {
+        if (baseID instanceof models.TBD) continue;
+
+        const base = this.lookup(baseID);
+        if (typeof base === 'undefined') {
+          logger.error('Reference to non-existing base: %s. ERROR_CODE:12002', baseID.fqn);
+          continue;
+        }
+
+      
+        if (base.hierarchy.length == 0 && base.basedOn.length > 0) {
+          this.expandHierarchy(base);
+        }
+
+        var manageValueInheritance = (parent, child) => {
+          if (child.inheritedFrom == null) {
+            if (parent.inheritedFrom) {
+              child.inheritedFrom = parent.inheritedFrom;
+            } else {
+              child.inheritedFrom = base.identifier;
+            };
+          }
+
+          if (child.constraints.length > 0 && child.identifier) {
+            for (const c of child.constraints) {
+              let matchedParentConstraint = parent.constraints.filter(p => p.equals(c)).pop();
+              if (matchedParentConstraint && matchedParentConstraint.lastModifiedBy) {
+                c.lastModifiedBy = matchedParentConstraint.lastModifiedBy;
+              } else {
+                c.lastModifiedBy = element.identifier;
+              };
+            }
+          } 
+        }
+
+        var manageCardHistory = (parent, child) => {
+          if (child.effectiveCard.equals(child.card)) {
+            return;
+          } else {
+            if (parent.constraintsFilter.own.card.constraints.length > 0) {
+              let oldCard = parent.constraintsFilter.own.card.constraints[0].card.clone();
+              oldCard.source = base.identifier;
+              if (parent.card.history.filter(h => h.source == oldCard.source).length == 0) { //if unique
+                parent.card.withHistory(oldCard);
+              }
+            } else {
+              let oldCard = parent.card.clone();
+              oldCard.source = base.identifier;
+              child.card.withHistory(oldCard)
+            }
+
+            if (parent.card.history) {
+              child.card.withHistories(parent.card.history);
```

Comment:
This adds the parent's histories _after_ the child's histories.  So the most recent history is at index `0` and they get older after that (and the very _original_ card is last in the array).  Just checking that this is the intentional order (as opposed to having index `0` represent the _original_ card).